### PR TITLE
Removing and replacing swagger-core with POROs

### LINKED ISSUE #20: Support for extensions
A colleague added an extension to our Swagger (using the @Extension annotation for Jersey) that looks like this:

    extensions = { @Extension(properties = { @ExtensionProperty(name = "x-cached", value = "true") }) })

This causes swagger-diff to fail:
```
   Hashie::CoercionError: Cannot coerce property "get" from Hash to Swagger::V2::Operation: The property 'x-cached' is not defined for Swagger::V2::Operation.
    /home/jenkins/.rvm/gems/ruby-2.1.2/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:46:in `rescue in set_value_with_coercion'
    /home/jenkins/.rvm/gems/ruby-2.1.2/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:34:in `set_value_with_coercion'
    /home/jenkins/.rvm/gems/ruby-2.1.2/gems/hashie-3.3.2/lib/hashie/extensions/indifferent_access.rb:107:in `indifferent_writer'
    /home/jenkins/.rvm/gems/ruby-2.1.2/gems/hashie-3.3.2/lib/hashie/dash.rb:168:in `block in initialize_attributes'
    /home/jenkins/.rvm/gems/ruby-2.1.2/gems/hashie-3.3.2/lib/hashie/dash.rb:167:in `each_pair'
    /home/jenkins/.rvm/gems/ruby-2.1.2/gems/hashie-3.3.2/lib/hashie/dash.rb:167:in `initialize_attributes'
    /home/jenkins/.rvm/gems/ruby-2.1.2/gems/hashie-3.3.2/lib/hashie/dash.rb:99:in `initialize'
    /home/jenkins/.rvm/gems/ruby-2.1.2/gems/swagger-core-0.2.3/lib/swagger/swagger_object.rb:16:in `initialize'
```

Is there any chance we could fix this?

Thanks,

David

### LINKED ISSUE #24: APIs.guru: showcase and test on real-life APIs
Hi @jeffreyc,

Maybe you would be interested in testing your tool against 200+ real-life Swagger files:
https://github.com/APIs-guru/api-models
They are validated by [sway](https://github.com/apigee-127/sway/tree/master/docs#swagger-validation) and could accessed through my [REST API](https://github.com/APIs-guru/api-models/blob/master/API.md)(no key needed).

You can also compare different versions of some popular API like Google Drive:
https://apis-guru.github.io/api-models/googleapis.com/drive/v1/swagger.yaml
https://apis-guru.github.io/api-models/googleapis.com/drive/v2/swagger.yaml
https://apis-guru.github.io/api-models/googleapis.com/drive/v3/swagger.yaml
IMHO it will help other developers to see all features of your tool.
All links support CORS so you can use it in online demo.
 
If you have any questions you can ask me directly through Skype(ivangon4arov), Hangout(ivan.goncharov.ua@gmail.com) or you can ask them on public chat: https://gitter.im/APIs-guru/api-models

### LINKED ISSUE #30: Unsupported response $ref
I have a swagger file containing `$ref`s to pre-defined responses. `swagger-diff` fails with an error:

My `resp-ref.yaml` file:

    ---
    swagger: '2.0'
    info:
      version: 0.0.0
      title: Simple API
    paths:
      /:
        get:
          responses:
            202:
              description: OK
            default:
              $ref: "#/responses/apiError"
    responses:
      apiError:
        description: "Any erroneous status code: 400 (parse error), 401 (auth error),\
          \ 402 (trial access), 403 (insufficient permissions), 404 (not found), 405 (unsupported\
          \ HTTP method), 415 (unsupported content type, 422 (validation error), 429 (request\
          \ rate limit exceeded), 500 (server error), 503 (maintenance)"
        schema:
          type: 'object'

When calling `swagger-diff resp-ref.yaml resp-ref.yaml` (I use same file for simplicity)
Error I am getting:

    /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:46:in `rescue in set_value_with_coercion': Cannot coerce property "responses" from Hash to {String=>Swagger::V2::Response}: The property '$ref' is not defined for Swagger::V2::Response. (Hashie::CoercionError)
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:34:in `set_value_with_coercion'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/indifferent_access.rb:107:in `indifferent_writer'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:168:in `block in initialize_attributes'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:167:in `each_pair'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:167:in `initialize_attributes'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:99:in `initialize'
        from /var/lib/gems/2.1.0/gems/swagger-core-0.2.3/lib/swagger/swagger_object.rb:16:in `initialize'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:68:in `new'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:68:in `block in coerce_or_init'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:35:in `call'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:35:in `set_value_with_coercion'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/indifferent_access.rb:107:in `indifferent_writer'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:168:in `block in initialize_attributes'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:167:in `each_pair'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:167:in `initialize_attributes'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:99:in `initialize'
        from /var/lib/gems/2.1.0/gems/swagger-core-0.2.3/lib/swagger/swagger_object.rb:16:in `initialize'
        from /var/lib/gems/2.1.0/gems/swagger-core-0.2.3/lib/swagger/v2/path.rb:20:in `initialize'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:68:in `new'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:68:in `block in coerce_or_init'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:40:in `call'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:40:in `block in set_value_with_coercion'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:40:in `each'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:40:in `map'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:40:in `set_value_with_coercion'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/indifferent_access.rb:107:in `indifferent_writer'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:168:in `block in initialize_attributes'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:167:in `each_pair'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:167:in `initialize_attributes'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:99:in `initialize'
        from /var/lib/gems/2.1.0/gems/swagger-core-0.2.3/lib/swagger/swagger_object.rb:16:in `initialize'
        from /var/lib/gems/2.1.0/gems/swagger-core-0.2.3/lib/swagger/api.rb:21:in `initialize'
        from /var/lib/gems/2.1.0/gems/swagger-core-0.2.3/lib/swagger/api.rb:9:in `new'
        from /var/lib/gems/2.1.0/gems/swagger-core-0.2.3/lib/swagger/api.rb:9:in `build_api'
        from /var/lib/gems/2.1.0/gems/swagger-core-0.2.3/lib/swagger.rb:24:in `build'
        from /var/lib/gems/2.1.0/gems/swagger-core-0.2.3/lib/swagger.rb:35:in `load'
        from /var/lib/gems/2.1.0/gems/swagger-diff-1.0.5/lib/swagger/diff/specification.rb:54:in `parse_swagger'
        from /var/lib/gems/2.1.0/gems/swagger-diff-1.0.5/lib/swagger/diff/specification.rb:6:in `initialize'
        from /var/lib/gems/2.1.0/gems/swagger-diff-1.0.5/lib/swagger/diff/diff.rb:5:in `new'
        from /var/lib/gems/2.1.0/gems/swagger-diff-1.0.5/lib/swagger/diff/diff.rb:5:in `initialize'
        from /var/lib/gems/2.1.0/gems/swagger-diff-1.0.5/exe/swagger-diff:6:in `new'
        from /var/lib/gems/2.1.0/gems/swagger-diff-1.0.5/exe/swagger-diff:6:in `<top (required)>'
        from /usr/local/bin/swagger-diff:23:in `load'
        from /usr/local/bin/swagger-diff:23:in `<main>'

### LINKED ISSUE #31: swagger-diff requires name for parameter $ref
I use parameter references in my swagger specs, and the diff fails because it expects parameter to have `name` attribute provided. 

As of [Swagger Specs](http://swagger.io/specification/#operationObject), parameter may contain a reference object, and reference object does not have `name` attribute.

My `param-ref.yaml` file 

    ---
    swagger: '2.0'
    info:
      version: 0.0.0
      title: Simple API
    paths:
      /:
        get:
          parameters:
            - $ref: "#/parameters/my-param"
          responses:
            200:
              description: OK
    parameters:
      my-param:
        name: "my-param"
        in: "query"
        description: "My test param"
        required: false
        type: "boolean"
    
When I call `swagger-tools param-ref.yaml param-ref.yaml` (same file just for simplicity) I get the below error. 

    /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:191:in `fail_property_required_error!': The property 'name' is required for Swagger::V2::Parameter. (ArgumentError)
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:183:in `assert_property_set!'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:178:in `block in assert_required_attributes_set!'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:177:in `each_key'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:177:in `assert_required_attributes_set!'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:100:in `initialize'
        from /var/lib/gems/2.1.0/gems/swagger-core-0.2.3/lib/swagger/swagger_object.rb:16:in `initialize'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:68:in `new'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:68:in `block in coerce_or_init'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:43:in `call'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:43:in `block in set_value_with_coercion'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:43:in `map'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:43:in `set_value_with_coercion'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/indifferent_access.rb:107:in `indifferent_writer'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:168:in `block in initialize_attributes'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:167:in `each_pair'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:167:in `initialize_attributes'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:99:in `initialize'
        from /var/lib/gems/2.1.0/gems/swagger-core-0.2.3/lib/swagger/swagger_object.rb:16:in `initialize'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:68:in `new'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:68:in `block in coerce_or_init'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:35:in `call'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:35:in `set_value_with_coercion'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/indifferent_access.rb:107:in `indifferent_writer'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:168:in `block in initialize_attributes'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:167:in `each_pair'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:167:in `initialize_attributes'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:99:in `initialize'
        from /var/lib/gems/2.1.0/gems/swagger-core-0.2.3/lib/swagger/swagger_object.rb:16:in `initialize'
        from /var/lib/gems/2.1.0/gems/swagger-core-0.2.3/lib/swagger/v2/path.rb:20:in `initialize'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:68:in `new'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:68:in `block in coerce_or_init'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:40:in `call'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:40:in `block in set_value_with_coercion'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:40:in `each'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:40:in `map'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/coercion.rb:40:in `set_value_with_coercion'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/extensions/indifferent_access.rb:107:in `indifferent_writer'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:168:in `block in initialize_attributes'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:167:in `each_pair'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:167:in `initialize_attributes'
        from /var/lib/gems/2.1.0/gems/hashie-3.3.2/lib/hashie/dash.rb:99:in `initialize'
        from /var/lib/gems/2.1.0/gems/swagger-core-0.2.3/lib/swagger/swagger_object.rb:16:in `initialize'
        from /var/lib/gems/2.1.0/gems/swagger-core-0.2.3/lib/swagger/api.rb:21:in `initialize'
        from /var/lib/gems/2.1.0/gems/swagger-core-0.2.3/lib/swagger/api.rb:9:in `new'
        from /var/lib/gems/2.1.0/gems/swagger-core-0.2.3/lib/swagger/api.rb:9:in `build_api'
        from /var/lib/gems/2.1.0/gems/swagger-core-0.2.3/lib/swagger.rb:24:in `build'
        from /var/lib/gems/2.1.0/gems/swagger-core-0.2.3/lib/swagger.rb:35:in `load'
        from /var/lib/gems/2.1.0/gems/swagger-diff-1.0.5/lib/swagger/diff/specification.rb:54:in `parse_swagger'
        from /var/lib/gems/2.1.0/gems/swagger-diff-1.0.5/lib/swagger/diff/specification.rb:6:in `initialize'
        from /var/lib/gems/2.1.0/gems/swagger-diff-1.0.5/lib/swagger/diff/diff.rb:5:in `new'
        from /var/lib/gems/2.1.0/gems/swagger-diff-1.0.5/lib/swagger/diff/diff.rb:5:in `initialize'
        from /var/lib/gems/2.1.0/gems/swagger-diff-1.0.5/exe/swagger-diff:6:in `new'
        from /var/lib/gems/2.1.0/gems/swagger-diff-1.0.5/exe/swagger-diff:6:in `<top (required)>'
        from /usr/local/bin/swagger-diff:23:in `load'
        from /usr/local/bin/swagger-diff:23:in `<main>'

### PR DESCRIPTION
**Pull request by jeffreyc**
@HawaiianSpork please review?

swagger-core did not handle the full Swagger specification. We weren't really benefiting from a parser (it did some massaging of operations but otherwise was a just enabling dot notation for hash access), so this replaces a 3rd-party parser with POROs. swagger-diff should now accept 100% of valid Swagger specifications.

Adding support for:
- parameter refs
- shared parameters
- response refs

Fixes #20
Fixes #30
Fixes #31
Closes #24

### CODE REVIEW

File: lib/swagger/diff/specification.rb [hunk 1]
```diff
@@ -44,36 +44,44 @@ def merge_refs!(h1, h2)
...
       def parsed_to_hash(parsed)
         ret = {}
-        parsed.operations.each do |endpoint|
-          ret["#{endpoint.verb} #{endpoint.path.gsub(/{.*?}/, '{}')}"] = endpoint
+        verbs = Set['get', 'put', 'post', 'delete', 'options', 'head', 'patch']
+        parsed['paths'].each do |path, items|
+          # TODO: this doesn't handle external definitions ($ref).
+          warn 'External definitions are not (yet) supported' if items.key?('$ref')
+          (verbs & items.keys).each do |verb|
+            if items['parameters']
```
Comment (HawaiianSpork) [inline lib/swagger/diff/specification.rb]: This code looks fine. Stylistically, you could probably extract the parameter rectifying code to its own method.  Also not sure if you care that this code changes the parsed object.

File: lib/swagger/diff/specification.rb [hunk 2]
```diff
@@ -84,9 +92,16 @@ def parsed_to_hash(parsed)
       # parameter definitions (i.e., all parameters, including nested
       # parameters, are included in a single set).
       def refs(ref, prefix = '')
+        defs = if ref[0..12] == '#/parameters/'
+                 @parsed['parameters']
+               elsif ref[0..11] == '#/responses/'
+                 @parsed['responses']
+               else
+                 @parsed['definitions']
```
Comment (HawaiianSpork) [inline lib/swagger/diff/specification.rb]: I would probably explicitly check for `'#/definitions/' to and throw an error if none of the cases are met.

Comment (jeffreyc) [inline lib/swagger/diff/specification.rb]: Added a warning and made sure it didn't raise an exception. For now, my philosophy is to try to be graceful with invalid/unexpected specifications.

File: lib/swagger/diff/specification.rb (hunk 1 above)
Comment (jeffreyc) [inline lib/swagger/diff/specification.rb]: Yeah, the mutation was a conscious decision. It's easier to treat the parameters as belonging to the operation, and not to have to check in two places.

**Comment by HawaiianSpork** [comment]
LGTM

**Pull request merged**